tsl/src/remote/copy_fetcher.c

```diff
@@ -338,85 +385,101 @@ copy_fetcher_complete(CopyFetcher *fetcher)
 			const int16 natts = copy_data_read_int16(&copy_data);
 			if (natts == -1)
 			{
-				fetcher->state.eof = true;
-				break;
-			}
+				/*
+				 * From the PostgreSQL (libpq) docs: The file trailer consists
+				 * of a 16-bit integer word containing -1. This is easily
+				 * distinguished from a tuple's field-count word. A reader
+				 * should report an error if a field-count word is neither -1
+				 * nor the expected number of columns. This provides an extra
+				 * check against somehow getting out of sync with the data.
+				 */
+				fetcher->file_trailer_received = true;
 
-			/*
-			 * There is also one case where no tupdesc attributes are retrieved.
-			 * This is when we do `select count(*) from t`, and
-			 * `enable_partitionwise_aggregate` is 0, so the data node queries
-			 * become `select null from ...` and we should get 1 NULL attribute
-			 * from COPY.
-			 */
-			int16 expected_natts = Max(1, retrieved_natts);
-			if (natts != expected_natts)
-			{
-				elog(ERROR,
-					 "wrong number of attributes for a COPY tuple: expected %d, got %d",
-					 expected_natts,
-					 natts);
+				/* Next PQgetCopyData() should return -1, indicating EOF and
+				 * that the remote side ended the copy. The final result
+				 * (PGRES_COMMAND_OK) should then be read with
+				 * PQgetResult(). */
 			}
-
-			Datum *values = &fetcher->batch_values[tupdesc_natts * row];
-			bool *nulls = &fetcher->batch_nulls[tupdesc_natts * row];
-			for (int i = 0; i < tupdesc_natts; i++)
+			else
 			{
-				nulls[i] = true;
-			}
+				Assert(!fetcher->file_trailer_received);
```

Comment:
`Assert(!eof)` also must hold, so the usage of `trailer`/`eof` is the same here again.